src/voxelalgorithms.cpp

```diff
@@ -885,6 +900,364 @@ void update_block_border_lighting(Map *map, MapBlock *block,
+ * the bottom of the voxel manipulator.
+ */
+void fill_with_sunlight(MMVManip *vm, INodeDefManager *ndef, v2s16 offset,
+	bool light[MAP_BLOCKSIZE][MAP_BLOCKSIZE])
+{
+	// Distance in array between two nodes on top of each other.
+	s16 step=vm->m_area.getExtent().X;
+	// Cache the ignore node.
+	MapNode ignore=MapNode(CONTENT_IGNORE);
+	// For each column of nodes:
+	for (s16 x = 0; x < MAP_BLOCKSIZE; x++) {
+		for (s16 z = 0; z < MAP_BLOCKSIZE; z++) {
+			// Position of the column on the map.
+			v2s16 realpos = offset + v2s16(x, z);
+			// Array indices in the voxel manipulator
+			s32 maxindex = vm->m_area.index(realpos.X, vm->m_area.MaxEdge.Y,
+				realpos.Y);
+			s32 minindex = vm->m_area.index(realpos.X, vm->m_area.MinEdge.Y,
+				realpos.Y);
+			// True if the current node has sunlight.
+			bool lig = light[x][z];
+			// For each node, downwards:
+			for (s32 i = maxindex; i >= minindex; i-=step) {
+				MapNode *n;
+				if (vm->m_flags[i] & VOXELFLAG_NO_DATA) {
+					n = &ignore;
+				} else {
+					n = &vm->m_data[i];
+				}
+				// Ignore IGNORE nodes, these are not generated yet.
+				if(n->getContent()==CONTENT_IGNORE)
+					continue;
+				const ContentFeatures &f = ndef->get(n->getContent());
+				if (lig && !f.sunlight_propagates) {
+					// Sunlight is stopped.
+					lig = false;
+				}
+				// Reset light
+				n->setLight(LIGHTBANK_DAY, lig ? 15 : 0, f);
+				n->setLight(LIGHTBANK_NIGHT, 0, f);
+			}
+			// Output outgoing light.
+			light[x][z] = lig;
+		}
+	}
+}
+
+/*!
+ * Returns incoming sunlight for one map block.
+ * If block above is not found, it is loaded.
+ *
+ * \param pos position of the map block that gets the sunlight.
+ * \param light incoming sunlight, light[x][z] is true if there
+ * is sunlight above the block at the given x-z relative
+ * node coordinates.
+ */
+void is_sunlight_above_block(ServerMap *map, mapblock_v3 pos,
+	INodeDefManager *ndef, bool light[MAP_BLOCKSIZE][MAP_BLOCKSIZE])
+{
+	mapblock_v3 source_block_pos = pos + v3s16(0, 1, 0);
+	// Get or load source block.
+	// It might take a while to load, but correcting incorrect
+	// sunlight may be even slower.
+	MapBlock *source_block = map->emergeBlock(source_block_pos, false);
+	// Trust only generated blocks.
+	if (source_block == NULL || source_block->isDummy()
+		|| !source_block->isGenerated()) {
+		// But if there is no block above, then use heuristics
+		bool sunlight = true;
+		MapBlock *node_block = map->getBlockNoCreateNoEx(pos);
+		if (node_block == NULL) {
+			// This should not happen.
+			sunlight = false;
+		} else {
+			sunlight = !node_block->getIsUnderground();
+		}
+		for (s16 x = 0; x < MAP_BLOCKSIZE; x++) {
+			for (s16 z = 0; z < MAP_BLOCKSIZE; z++) {
+				light[x][z] = sunlight;
+			}
+		}
+	} else {
+		// Dummy boolean, the position is valid.
+		bool is_valid_position;
+		// For each column:
+		for (s16 x = 0; x < MAP_BLOCKSIZE; x++) {
+			for (s16 z = 0; z < MAP_BLOCKSIZE; z++) {
```

Comment:
Either way is fine. Occasionally it is best to do ZYX or ZX order but only when there is a reason such as incrementing indexes by +1 along an X-row.